pkg/templates/targetport/template.go

```diff
@@ -0,0 +1,102 @@
+package targetport
+
+import (
+	"fmt"
+
+	"golang.stackrox.io/kube-linter/pkg/check"
+	"golang.stackrox.io/kube-linter/pkg/config"
+	"golang.stackrox.io/kube-linter/pkg/diagnostic"
+	"golang.stackrox.io/kube-linter/pkg/extract"
+	"golang.stackrox.io/kube-linter/pkg/extract/customtypes"
+	"golang.stackrox.io/kube-linter/pkg/lintcontext"
+	"golang.stackrox.io/kube-linter/pkg/objectkinds"
+	"golang.stackrox.io/kube-linter/pkg/templates"
+	"golang.stackrox.io/kube-linter/pkg/templates/targetport/internal/params"
+	coreV1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	k8sValidation "k8s.io/apimachinery/pkg/util/validation"
+)
+
+const (
+	templateKey = "target-port"
+)
+
+func init() {
+	templates.Register(check.Template{
+		HumanName:   "Target Port",
+		Key:         templateKey,
+		Description: "Flag containers and services using not allowed port names or numbers",
+		SupportedObjectKinds: config.ObjectKindsDesc{
+			ObjectKinds: []string{objectkinds.DeploymentLike, objectkinds.Service},
+		},
+		Parameters:             params.ParamDescs,
+		ParseAndValidateParams: params.ParseAndValidate,
+		Instantiate: params.WrapInstantiateFunc(func(p params.Params) (check.Func, error) {
+			return func(_ lintcontext.LintContext, object lintcontext.Object) []diagnostic.Diagnostic {
+				if !p.TargetPort {
+					return nil
+				}
+
+				podSpec, foundPodSpec := extract.PodSpec(object.K8sObject)
+				if foundPodSpec {
+					return findPodPorts(&podSpec)
+				}
+
+				service, foundService := object.K8sObject.(*coreV1.Service)
+				if foundService {
+					return findServicePorts(service)
+				}
+
+				return nil
+			}, nil
+
+		}),
+	})
+}
+
+func findPodPorts(podSpec *customtypes.PodSpec) []diagnostic.Diagnostic {
+	var results []diagnostic.Diagnostic
+
+	containers := podSpec.AllContainers()
+	for _, container := range containers {
+		for _, port := range container.Ports {
```

Comment:
I think we should skip if `port.Name` is empty as we have no way to detect if it was empty or null.

Reply:
@janisz, I have pushed the changes. We have a similar problem with `Service` if `targetPort` is not provided. It's the same as `int 0`. I have adjusted that too. I have also checked the outcome of the dry-run, and `targetPort: 0` is valid. 😞